src/tree-set.satyg

```diff
@@ -0,0 +1,238 @@
+@stage: persistent
+@import: base
+@import: ord
+@import: fn
+@import: option-ext
+@import: int
+@import: stream
+
+module TreeSet : sig
+  % Immutable set.
+  type 'a t
+
+  % Creates an empty set.
+  val empty : 'a t
+
+  % Returns the number of elements in the given set.
+  val size : 'a t -> int
+
+  % Returns [true] if the given set is empty.
+  val is-empty : 'a t -> bool
+
+  % [x |> mem ord s] returns [true] if [s] contains [x].
+  val mem : ('a Ord.t) implicit -> 'a t -> 'a -> bool
+
+  % [contains ord x s] is the same as [mem ord s x].
+  val contains : ('a Ord.t) implicit -> 'a -> 'a t -> bool
+
+  % [get ord x s] Returns an element that is equal to [x] with respect to [ord].
+  val get : ('a Ord.t) implicit -> 'a -> 'a t -> 'a option
+
+  % Adds a value to the given set. If the set contains a value that is equal to the new value,
+  % the old value is replaced with the new value. If you need to get the old value, query [get] first.
+  val insert : ('a Ord.t) implicit -> 'a -> 'a t -> 'a t
+
+  % Removes a value from the given set.
+  val remove : ('a Ord.t) implicit -> 'a -> 'a t -> 'a t
+
+  % Constructs a set from a list. If the list contains two equal values, the preceding one is preferred.
+  val of-list : ('a Ord.t) implicit -> 'a list -> 'a t
+
+  val to-list : 'a t -> 'a list
+
+  val to-stream : 'a t -> 'a Stream.t
+end = struct
+
+  % AVL-tree's invariant:
+  % (INV) for each node n, |height(n#left) - height(n#right)| < 2
+
+  % AVL tree
+  type 'a node = Node of ((|
+    height : int;
+    value : 'a;
+    left : 'a node;
+    right : 'a node;
+  |)) option
+
+  type 'a t = Tree of (|
+    root : 'a node;
+    size : int;
+  |)
+
+  let empty-node = Node(None)
+
+  let empty = Tree (|
+    root = empty-node;
+    size = 0;
+  |)
+
+  let size (Tree t) = t#size
+
+  let is-empty-node (Node(n)) = Option.is-none n
+
+  let is-empty t = size t == 0
+
+  let get ord x (Tree t) =
+    t#root |> Fn.fix (fun rec (Node(node)) -> (
+      match node with
+      | None -> None
+      | Some(node) ->
+        (match Ord.compare ord x node#value with
+        | Eq -> Some(node#value)
+        | Lt -> rec node#left
+        | Gt -> rec node#right)
+    ))
+
+  let mem ord s x = Option.is-some (get ord x s)
+
+  let contains ord x s = mem ord s x
+
+  let height (Node n) =
+    match n with
+    | None -> 0
+    | Some(node) -> node#height
+
+  let make-node left right value =
+    Node (Some (|
+      height = 1 + Int.max (height left) (height right);
+      left = left;
+      right = right;
+      value = value;
+    |))
+
+  % If both l and r satisfy (INV) and |height(l) - height(r)| <= 2, so does [balance l r v].
+  let balance l r v =
+    let hl = height l in
+    let hr = height r in
+    if Int.(abs (hl - hr) < 2) then
```

Comment:
More like a question: How `Int.(abs hl - hr < 2)` work? Is it the same as `open Int in (abs hl - hr < 2)`?

Reply:
Correct. They have precisely the same effect.